msg/templates/urtps/microRTPS_timesync.cpp.em

```diff
@@ -0,0 +1,234 @@
+@[    else]@
+			timesync msg = newTimesyncMsg();
+@[    end if]@
+@[end if]@
+
+			_timesync_pub.publish(&msg);
+
+			std::this_thread::sleep_for(std::chrono::milliseconds(100));
+		}
+	};
+	_request_stop = false;
+	_send_timesync_thread.reset(new std::thread(run));
+}
+
+void TimeSync::stop() {
+	_request_stop = true;
+	if (_send_timesync_thread && _send_timesync_thread->joinable()) _send_timesync_thread->join();
+	_send_timesync_thread.reset();
+}
+
+void TimeSync::reset() {
+	_num_samples = 0;
+	_request_reset_counter = 0;
+}
+
+bool TimeSync::addMeasurement(int64_t local_t1_ns, int64_t remote_t2_ns, int64_t local_t3_ns) {
+	int64_t rtti = local_t3_ns - local_t1_ns;
+
+	// assume rtti is evenly split both directions
+	int64_t remote_t3_ns = remote_t2_ns + rtti / 2ll;
+
+	int64_t measurement_offset = remote_t3_ns - local_t3_ns;
+
+	if (_request_reset_counter > REQUEST_RESET_COUNTER_THRESHOLD) {
+		reset();
+		std::cout << std::endl << "Timesync clock changed, resetting" << std::endl;
+	}
+
+	if (_num_samples >= WINDOW_SIZE) {
+		if (std::abs(measurement_offset - _offset_ns.load()) > TRIGGER_RESET_THRESHOLD_NS) {
+			_request_reset_counter++;
+			std::cout << std::endl << "Timesync offset outlier, discarding" << std::endl;
+			return false;
+		} else {
+			_request_reset_counter = 0;
+		}
+	}
+
+	if (_num_samples == 0) {
+		updateOffset(measurement_offset);
+		_skew_ns_per_sync = 0;
+	}
+
+	// ignore if rtti > 10ms
+	if (rtti > 15ll * 1000ll * 1000ll) {
+		std::cout << std::endl
+			  << "RTTI too high for timesync: " << rtti / (1000ll * 1000ll) << "ms" << std::endl;
+		return false;
+	}
+
+	double schedule = std::min((double)_num_samples / WINDOW_SIZE, 1.);
+	double alpha = ALPHA_INITIAL * (1. - schedule) + ALPHA_FINAL * schedule;
+	double beta = BETA_INTIIAL * (1. - schedule) + BETA_FINAL * schedule;
+
+	int64_t offset_prev = _offset_ns.load();
+	updateOffset(static_cast<int64_t>((_skew_ns_per_sync + _offset_ns.load()) * (1. - alpha) + measurement_offset * alpha));
+	_skew_ns_per_sync = static_cast<int64_t>(beta * (_offset_ns.load() - offset_prev) + (1. - beta) * _skew_ns_per_sync);
+
+	_num_samples++;
+
+	return true;
+}
+
+@[if 1.5 <= fastrtpsgen_version <= 1.7]@
+@[    if ros2_distro]@
+void TimeSync::processTimesyncMsg(@(package)::msg::dds_::Timesync_* msg) {
+@[    else]@
+void TimeSync::processTimesyncMsg(timesync_* msg) {
+@[    end if]@
+@[else]@
+@[    if ros2_distro]@
+void TimeSync::processTimesyncMsg(@(package)::msg::Timesync* msg) {
+@[    else]@
+void TimeSync::processTimesyncMsg(timesync* msg) {
+@[    end if]@
+@[end if]@
+	if (msg->sys_id() == 1 && msg->seq() != _last_remote_msg_seq) {
+                if (msg->tc1() > 0) {
+                        _last_remote_msg_seq = msg->seq();
+
+                        if (!addMeasurement(msg->ts1(), msg->tc1(), getMonoRawTimeNSec())) {
+                                std::cerr << "Offset not updated" << std::endl;
+                        }
+
+	        } else if (msg->tc1() == 0) {
```

Comment:
You don't need to handle the `else` here, it is already done in the `send`. Unless I misunderstand the structure, and this same code gets executed on both sides of the link?

Reply:
This one is here but for the future case when it's the FMU starting the timesync. It's currently not being called.